sklearn/ensemble/tests/test_forest.py

```diff
@@ -1377,3 +1377,22 @@ def test_little_tree_with_small_max_samples(ForestClass):
 
     msg = "Tree without `max_samples` restriction should have more nodes"
     assert tree1.node_count > tree2.node_count, msg
+
+
+@pytest.mark.parametrize('Forest', FOREST_REGRESSORS)
+def test_mse_criterion_object_segfault_smoke_test(Forest):
+    # Ensure that we can pass a mutable criterion while using parallel fit
+    # Non-regression test for:
+    # https://github.com/scikit-learn/scikit-learn/issues/12623
+    from sklearn.tree._classes import CRITERIA_REG
+
+    X = np.random.random((1000, 3))
+    y = np.random.random((1000, 1))
+
+    n_samples, n_outputs = y.shape
+    mse_criterion = CRITERIA_REG['mse'](n_outputs, n_samples)
+    est = FOREST_REGRESSORS[Forest](
+        n_estimators=2, n_jobs=2, criterion=mse_criterion
```

Comment:
If `FOREST_REGRESSORS[Forest]` knows that it is going to mutate a hyper-parameter such as `criterion`, I think `est.fit` should have the responsibility of copying `criterion`.

WDYT @glemaitre ?

Reply:
Hi @thomasjpfan - I'm resuming this for Alex and was wondering what you think we should do next.

If I remember correctly, I think since the underlying issue is specific to a certain presumption of thread-safety when instantiating estimators we all settled on fixing it within [`_make_estimator(...)`](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/ensemble/_base.py#L144) as suggested by the comments within that method.

Reply:
I was thinking the deepcopy may  be costly for certain objects and a more targeted solution would be to copy in `BaseDecisionTree`:

https://github.com/scikit-learn/scikit-learn/blob/94abe05b4b96de2ca30d998fb9adb2fbd3eb1bde/sklearn/tree/_classes.py#L344-L351

```py
if not isinstance(criterion, Criterion): 
    ...
else:
    criterion = copy.deepcopy(criterion)
```

Reply:
This would work for us, although I'm not familiar enough with the rest of the ensemble implementation to know if it categorically covers all potential multi-threading issues.

I will update this fix and create a PR asap. Thanks again for looking into this.